datalad_next/archive_operations/__init__.py

```diff
@@ -0,0 +1,65 @@
+"""Archive operation handlers"""
+
+# allow for |-type UnionType declarations
+from __future__ import annotations
+
+import logging
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from contextlib import contextmanager
+from typing import (
+    Any,
+    Generator,
+    IO,
+)
+
+import datalad
+
+from datalad_next.config import ConfigManager
+from datalad_next.iter_collections.utils import FileSystemItem
+
+
+lgr = logging.getLogger('datalad.ext.next.archive_operations')
+
+
+class ArchiveOperations(ABC):
+    """
+    """
+    def __init__(self, location: Any, *, cfg: ConfigManager | None = None):
+        """
+        Parameters
+        ----------
+        location:
+          Archive location identifier (path, URL, etc.) understood by a
+          particular archive handler.
+        cfg: ConfigManager, optional
+          A config manager instance that implementations will consult for
+          any configuration items they may support.
+        """
+        self._cfg = cfg
+
+    @property
+    def cfg(self) -> ConfigManager:
+        if self._cfg is None:
+            self._cfg = datalad.cfg
+        return self._cfg
+
+    @contextmanager
+    @abstractmethod
+    def open(self, item: Any) -> IO:
+        """
+        """
+        raise NotImplementedError
+
+    def close(self) -> None:
+        pass
+
+    @abstractmethod
+    def __contains__(self, item: Any) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __iter__(self) -> Generator[FileSystemItem, None, None]:
```

Comment:
I am not convinced that this is a good choice, but maybe it is. We are operating under the assumption that all iterator will produce that. I have no good feeling whether that is valid.

Reply:
I share a "slightly weird" feeling. It is better than `TarfileItem`. I think we have to answer a general question:

- Should the yield-type of `ArchiveOperations.__iter__` and of `iter_<x>` be in the same class hierarchy?

For example:

```python
# The name below is stupid and just used to illustrate the point 
@dataclass
class CollectionOrArchiveItem:
    name: PurePosixPath




@dataclass
class FilesystemItem(CollectionOrArchiveItem):
    type: FileSystemItemType
    size: int
    mtime: float | None = None
    mode: int | None = None
    uid: int | None = None
    gid: int | None = None
    link_target: PurePath | None = None
```


Pro: it would make sense because both are members of some of set that distinguishes its elements by their name (although the elements usually have additional properties, the names are the unique distinction criteria). So there could be a common base-class for those types that basically contains a name.

Con: there are distinctions between the names in archives and names in collections. Names in archives are specified by the archive specs and should be identical across platforms. Names in collections might not be identical across platforms, i.e. Windows-dirs vs. Unix-dirs, unless they are represented as strings or a sequence of bytes (which we are currently doing)

Reply:
I arrived at similar conclusions. Please check https://github.com/datalad/datalad-next/issues/350 and specifically https://github.com/datalad/datalad-next/pull/361/commits/63d418e69ee1883c23185ca012e36cd8bc071256 for my approach from 1-2 weeks ago.